scipy/constants/_constants.py

```diff
@@ -13,7 +13,7 @@
 from typing import TYPE_CHECKING, Any
 
 from ._codata import value as _cd
-import numpy as _np
+import numpy as _np # noqa: ICN001
```

Comment:
I think this looks right to me. From PEP8,

> `_single_leading_underscore`: weak “internal use” indicator. E.g. `from M import *` does not import objects whose names start with an underscore.

So I think there is a case to be made that `import numpy as _np` is actually _more_ correct.

However, it clearly isn't the convention. There's probably some historical discussion about this. I think `noqa` is fine for now, maybe others will have an opinion.

Reply:
[PEP8](https://peps.python.org/pep-0008/#public-and-internal-interfaces) actually says:

> Even with __all__ set appropriately, internal interfaces (packages, modules, classes, functions, attributes or other names) should still be prefixed with a single leading underscore.

An interface is also considered internal if any containing namespace (package, module or class) is considered internal.

In which case I suggest adding `"numpy" = "_np"` to `[lint.flake8-import-conventions.extend-aliases]`. Again, the opinion of others may differ.

I wonder whether ruff should allow leading underscores by default.

Reply:
I'd just fix it. There is no reason here to not do `import numpy as np` and using `_np` looks weird; this file already starts with an underscore and has an `__all__` dict, so there shouldn't even be a worry about exposing `np` in the public `constants` namespace.

Reply:
Yeah fair enough, no reason for this to be inconsistent with the vast majority of the codebase.

> this file already starts with an underscore

What about the other files? I assume we don't really want to be able to do

```
In [1]: from scipy.linalg.blas import np
```

Reply:
`__all__` is mostly taking care of this indeed.

You can explicitly do `from scipy.linalg.blas import np`, but you can also do that for anything else in `__init__.py` files. When writing a new package, best practice is probably to have nothing in `__init__.py` files except `from _xxx import yyy`; that is about the only thing that's water-tight. But very few people do it like that, and it's not a big deal in practice.

Reply:
Makes sense. Worth doing the refactor if someone ever complains I guess, but for now no point to go the extra mile here.